src/subnet_rental_canister/src/canister.rs

```diff
@@ -69,6 +71,85 @@ fn start_timers() {
+            println!(
+                "Busy processing another request. Skipping cycles burn for subnet {}",
+                rental_agreement.subnet_id
+            );
+            continue;
+        };
+
+        // Because of the copy in the loop head, by the time we execute here, the rental agreement might not exist anymore.
+        let Some(rental_agreement) = get_rental_agreement(&rental_agreement.subnet_id) else {
+            println!(
+                "Rental agreement for subnet {} no longer exists. Skipping cycles burn.",
+                rental_agreement.subnet_id
+            );
+            continue;
+        };
+
+        // Now we are sure the rental agreement still exists and only we can modify it.
+
+        let total_cycles_burned = rental_agreement.total_cycles_burned;
+        let total_cycles_created = rental_agreement.total_cycles_created;
+        let total_cycles_remaining = total_cycles_created.saturating_sub(total_cycles_burned);
+        let paid_until_nanos = rental_agreement.paid_until_nanos;
+        let now_nanos = ic_cdk::api::time();
+
+        // No more cycles left to burn.
+        if total_cycles_remaining == 0 {
+            continue;
+        }
+
+        // The rental agreement is not paid for anymore.
+        if now_nanos >= paid_until_nanos {
+            // Burn all remaining cycles.
+            println!(
+                "Burning all remaining cycles for subnet {} (subnet is no longer paid for)",
+                rental_agreement.subnet_id
+            );
+            let burned = ic_cdk::api::cycles_burn(total_cycles_remaining);
+            if let Err(e) = update_rental_agreement(rental_agreement.subnet_id, |mut agreement| {
+                agreement.total_cycles_burned =
+                    agreement.total_cycles_burned.saturating_add(burned);
+                agreement
+            }) {
+                println!(
+                    "Failed to update rental agreement for subnet {}: {}. Skipping.",
+                    rental_agreement.subnet_id, e
+                );
+            }
+            continue;
+        }
+
+        // The rental agreement is still paid for and has some cycles left to burn.
+        let time_remaining_nanos = paid_until_nanos - now_nanos; // guaranteed to be positive due to above check
+        let time_remaining_seconds = time_remaining_nanos / BILLION;
+        let amount_to_burn_per_minute = (total_cycles_remaining / time_remaining_seconds as u128)
+            * CYCLES_BURN_INTERVAL_SECONDS as u128;
+
+        let burned =
+            ic_cdk::api::cycles_burn(min(amount_to_burn_per_minute, total_cycles_remaining)); // don't burn more than we have
+        if let Err(e) = update_rental_agreement(rental_agreement.subnet_id, |mut agreement| {
+            agreement.total_cycles_burned = agreement.total_cycles_burned.saturating_add(burned);
+            agreement
+        }) {
```

Comment:
No change required.

I think we should really avoid multi-line if conditions. IMHO, using "dummy" variables can be highly intelligent!